File: big_data_learning/spark-learning/com/shawn/study/spark/core/wordcount/word-count.py
from pyspark import SparkConf, SparkContext

data = list(['Hello World', 'Hello Spark', 'Hello Hadoop', 'Spark Learning'])


# wordCount by reduceByKey
def word_count_by_reduce_by_key(spark_context: SparkContext):
    print("word_count_by_reduce_by_key: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda line: line.split(" "))
          .map(lambda word: (word, 1))
          .reduceByKey(lambda a, b: a + b)
          .collect())


# wordCount by groupBy
def word_count_by_group_by(spark_context: SparkContext):
    print("word_count_by_group_by: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .groupBy(lambda word: word)
          .map(lambda x: (x[0], len(list(x[1]))))
          .collect())


# word count by group by key
def word_count_by_group_by_key(spark_context: SparkContext):
    print("word_count_by_group_by_key: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .map(lambda word: (word, 1))
          .groupByKey()
          .mapValues(lambda x: len(x))
          .collect())


# word count by aggregate_key
def word_count_by_aggregate_key(spark_context: SparkContext):
    print("word_count_by_aggregate_key: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .map(lambda word: (word, 1))
          .aggregateByKey(0, lambda k1, k2: k1 + k2, lambda v1, v2: v1 + v2)
          .collect())


def word_count_by_fold_key(spark_context: SparkContext):
    print("word_count_by_fold_key: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .map(lambda word: (word, 1))
          .foldByKey(0, lambda v1, v2: v1 + v2)
          .collect())


def word_count_by_combine_key(spark_context: SparkContext):
    print("word_count_by_combine_key: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .map(lambda word: (word, 1))
          .combineByKey(lambda v: v, lambda x, y: x + y, lambda x, y: x + y)
          .collect())


def word_count_by_count_key(spark_context: SparkContext):
    print("word_count_by_count_key: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .map(lambda word: (word, 1))
          .countByKey())


def word_count_by_count_by_value(spark_context: SparkContext):
    print("word_count_by_count_by_value: ")
    print(spark_context.parallelize(data)
          .flatMap(lambda word: word.split(" "))
          .countByValue())


if __name__ == '__main__':
    conf = SparkConf().setMaster("local").setAppName('spark-learning')
    sc = SparkContext(conf=conf)

    word_count_by_reduce_by_key(spark_context=sc)
    word_count_by_group_by(spark_context=sc)
    word_count_by_group_by_key(spark_context=sc)
    word_count_by_aggregate_key(spark_context=sc)
    word_count_by_fold_key(spark_context=sc)
    word_count_by_combine_key(spark_context=sc)
    word_count_by_count_key(spark_context=sc)
    word_count_by_count_by_value(spark_context=sc)

    # done!
    sc.stop()

File: big_data_learning/spark/code/scala/src/main/java/com/shawn/study/big/data/spark/scala/wc/WordCount.scala
package com.shawn.study.big.data.spark.scala.wc

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

import scala.collection.mutable

object WordCount {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("wordCount");
    val sc = new SparkContext(conf)
    val path = "../../data/wc/word_count_example"
    val rdd = sc.textFile(path)
    wc1(rdd)
    wc2(rdd)
    wc3(rdd)
    wc4(rdd)
    wc5(rdd)
    wc6(rdd)
    wc7(rdd)
    wc8(rdd)
    wc9(rdd)
    wc10(rdd)
    sc.stop()
  }

  // flatMap + groupBy + map
  def wc1(rdd: RDD[String]): Unit = {
    rdd.flatMap(_.split(" ")).groupBy(word => word).map({
      case (word, list) =>
        (word, list.size)
    }).collect().foreach(println)
  }

  // flatMap + groupBy + mapValues
  def wc2(rdd: RDD[String]): Unit = {
    val group = rdd.flatMap(_.split(" ")).groupBy(word => word)
    group.mapValues(iter => iter.size).collect().foreach(println)
  }

  // flatMap + map + groupByKey + mapValues
  def wc3(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map((_, 1))
    val groupByKey = map.groupByKey()
    val mapValues = groupByKey.mapValues(iter => iter.size)
    mapValues.collect().foreach(println)
  }

  // flatMap + map + reduceByKey
  def wc4(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map((_, 1))
    val reduceByKey = map.reduceByKey(_ + _)
    reduceByKey.collect().foreach(println)
  }

  // flatMap + map + aggregateByKey
  def wc5(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map((_, 1))
    val aggregateByKey = map.aggregateByKey(0)(_ + _, _ + _)
    aggregateByKey.collect().foreach(println)
  }

  // flatMap + map + foldByKey
  def wc6(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map((_, 1))
    val foldByKey = map.foldByKey(0)(_ + _)
    foldByKey.collect().foreach(println)
  }

  // flatMap + map + combineByKey
  def wc7(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map((_, 1))
    val combineByKey = map.combineByKey(v => v, (x: Int, y) => x + y, (x: Int, y: Int) => x + y)
    combineByKey.collect().foreach(println)
  }

  // flatMap + map + countByKey
  def wc8(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map((_, 1))
    val countByKey = map.countByKey()
    countByKey.keys.foreach(key => println("(" + key + ", " + countByKey(key) + ")"))
  }

  // flatMap + countByValue
  def wc9(rdd: RDD[String]): Unit = {
    val flatMap = rdd.flatMap(_.split(" "))
    val countByValue = flatMap.countByValue()
    countByValue.keys.foreach(key => println("(" + key + ", " + countByValue(key) + ")"))
  }

  // flatMap + map + reduce
  def wc10(rdd: RDD[String]): Unit = {
    val map = rdd.flatMap(_.split(" ")).map(word => {
      mutable.Map[String, Long]((word, 1))
    })
    val value = map.reduce((map1, map2) => {
      map2.foreach {
        case (word, count) =>
          val newCount = map1.getOrElse(word, 0L) + count
          map1.update(word, newCount)
      }
      map1
    })
    println(value)
  }
}
